src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.LazySyntaxTree.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    public partial class CSharpSyntaxTree
+    {
+        private sealed class LazySyntaxTree : CSharpSyntaxTree
+        {
+            private readonly SourceText _text;
+            private readonly CSharpParseOptions _options;
+            private readonly string _path;
+            private readonly ImmutableDictionary<string, ReportDiagnostic> _diagnosticOptions;
+            private CSharpSyntaxNode? _lazyRoot;
+
+            internal LazySyntaxTree(
```

Comment:
I'm somewhat on the fence about introducing a new type of SyntaxTree but I agree it makes sense given the API we have today. 

I actually think we should avoid returning realized `SyntaxTree` instances from the generator at all, and instead have a method that does it lazily. Given that that is a breaking change however, lets punt it down the road, and take this approach for now.

Reply:
Agreed; I think this is the best approach _for now_, and we can consider using the next major version to just break the API.